src/v/cluster/rm_stm.cc

```diff
@@ -730,31 +730,45 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate(
   model::batch_identity bid,
   model::record_batch_reader b,
   raft::replicate_options opts) {
-    return _state_lock.hold_read_lock().then(
-      [this, bid, opts, b = std::move(b)](
-        ss::basic_rwlock<>::holder unit) mutable {
-          if (bid.is_transactional) {
-              auto pid = bid.pid.get_id();
-              return get_tx_lock(pid)
-                ->with([this, bid, b = std::move(b)]() mutable {
-                    return replicate_tx(bid, std::move(b));
-                })
-                .finally([u = std::move(unit)] {});
-          } else if (bid.has_idempotent() && bid.first_seq <= bid.last_seq) {
-              return replicate_seq(bid, std::move(b), opts)
-                .finally([u = std::move(unit)] {});
-          }
-
-          return sync(_sync_timeout)
-            .then([this, opts, b = std::move(b)](bool is_synced) mutable {
-                if (!is_synced) {
-                    return ss::make_ready_future<
-                      result<raft::replicate_result>>(errc::not_leader);
-                }
-                return _c->replicate(std::move(b), opts);
-            })
-            .finally([u = std::move(unit)] {});
-      });
+    return _state_lock.hold_read_lock().then([this,
+                                              bid,
+                                              opts,
+                                              b = std::move(b)](
+                                               ss::basic_rwlock<>::holder
+                                                 unit) mutable {
+        if (bid.is_transactional) {
+            auto pid = bid.pid.get_id();
+            return get_tx_lock(pid)
+              ->with([this, bid, b = std::move(b)]() mutable {
+                  return replicate_tx(bid, std::move(b));
+              })
+              .finally([u = std::move(unit)] {});
+        } else if (bid.has_idempotent() && bid.first_seq <= bid.last_seq) {
+            request_id rid{.pid = bid.pid, .seq = bid.first_seq};
+            auto request_lock = get_request_lock(rid);
+
+            return request_lock->with(
+              [this, request_lock, bid, rid, opts, b = std::move(b)]() mutable {
+                  return replicate_seq(bid, std::move(b), opts)
+                    .finally([this, request_lock, rid]() {
+                        if (request_lock->waiters() == 1) {
+                            // this fiber is the last holder of the lock
+                            _request_locks.erase(rid);
+                        }
+                    });
+              });
```

Comment:
I wonder if it's worth simplifying the code here with a helper:
```c++
    template<typename Func>
    auto with_request_lock(request_id rid, Func&& func) {
        // Maybe also inline get_request_lock()
        auto r_lock = get_request_lock(rid);
        return r_lock->with(std::forward<Func>(func))
          .finally([this, r_lock, rid]() {
              if (r_lock->waiters() == 1) {
                  // this fiber is the last holder of the lock
                  _request_locks.erase(rid);
              }
          });
    }
```
And then:
```c++
              return with_request_lock(
                rid, [this, bid, opts, b = std::move(b)]() mutable {
                    return replicate_seq(bid, std::move(b), opts);
                });
```

Reply:
done